rust/src/shards/gui/widgets/file_dialog.rs

```diff
@@ -0,0 +1,147 @@
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::NONE_TYPES;
+use crate::types::STRING_TYPES;
+
+lazy_static! {
+  static ref FILEDIALOG_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Folder"),
+      shccstr!("To select a folder instead of a file."),
+      BOOL_VAR_OR_NONE_SLICE,
+    )
+      .into(),
+  ];
+}
+
+impl Default for FileDialog {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      folder: ParamVar::new(false.into()),
+    }
+  }
+}
+
+impl Shard for FileDialog {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.FileDialog")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.FileDialog-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.FileDialog"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &NONE_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &STRING_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&FILEDIALOG_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.folder.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.folder.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.folder.warmup(ctx);
+    self.parents.warmup(ctx);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.parents.cleanup();
+    self.folder.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if let Some(ui) = util::get_current_parent(self.parents.get())? {
+      self.activate_ui(ui, context, input)
+    } else {
+      Err("No UI parent")
+    }
+  }
+}
+
+impl FileDialog {
+  // FIXME rfd is supported on Linux but requires linking with GTK3
+  #[cfg(any(target_os = "windows", target_os = "macos"))]
+  fn activate_ui(&mut self, ui: &mut egui::Ui, _context: &Context, _input: &Var) -> Result<Var, &str> {
+    // Note: technically we don't need egui when using rfd, but we still keep the dependency for consistency with the fallback
+    let path = if self.folder.get().try_into()? {
+      rfd::FileDialog::new().pick_folder()
+    } else {
+      rfd::FileDialog::new().pick_file()
+    };
+    let output = if let Some(path) = path {
+      let tmp = Var::ephemeral_string(path.display().to_string().as_str());
+      let mut output = Var::default();
+      cloneVar(&mut output, &tmp);

```

Comment:
Avoid using cloneVar wherever possible, shards shouldn't output cloned vars as this will leak.
If you need storage it should be in one of the rust helper structs like ClonedVar and stored on the shards and return a reference.

Reply:
For string it is the only way as far as I know: the ephemeral string cannot be kept around.

Reply:
this is a leak, you need to store output as `ClonedVar` in self and stop using `cloneVar`

Reply:
The thing is it also needs to be releaseVar'd, but ClonedVar takes care of that since it implements drop.
By convention shards return values should be owned by the shard returning it, so store the result in a ClonedVar member (might need option<> around it) and return that converted to SHVar